File: docker/Dockerfile.swift
# Cross-platform Swift implementation. Socket layer supports Windows (WinSDK),
# Linux (Glibc), and macOS (Darwin) via conditional compilation. Build with:
#   - Windows: any Swift toolchain that ships WinSDK
#   - Linux:   swift:6.3 (used here for Docker)
#   - macOS:   swift build on Darwin
FROM swift:6.3 AS build-base
COPY hello-websocket-swift /app/hello-websocket-swift
WORKDIR /app/hello-websocket-swift
RUN swift build -c release

FROM swift:6.3-slim AS server
RUN useradd --system --create-home --shell /usr/sbin/nologin app
WORKDIR /app
COPY --from=build-base --chown=app:app /app/hello-websocket-swift/.build/release/ws-server /app/
USER app
ENTRYPOINT ["./ws-server"]

FROM swift:6.3-slim AS client
RUN useradd --system --create-home --shell /usr/sbin/nologin app
WORKDIR /app
COPY --from=build-base --chown=app:app /app/hello-websocket-swift/.build/release/ws-client /app/
USER app
ENTRYPOINT ["./ws-client"]

File: docker/Dockerfile.go
# hello-websocket-go/go.mod declares `go 1.23`, the minimum toolchain
# version this module requires (per Go's module compatibility contract,
# https://go.dev/doc/toolchain). Building with a newer toolchain such as
# 1.26 here is expected and safe; it is not a version mismatch to "fix" by
# keeping this pinned to 1.23.
FROM golang:1.26-alpine AS build-base
RUN apk add --update git && rm -rf /var/cache/apk/*
ENV GOPROXY=https://goproxy.cn,direct
ENV GO111MODULE=on

COPY hello-websocket-go /app/hello-websocket-go
WORKDIR /app/hello-websocket-go
RUN go mod download
RUN go build -o ws_server server/main.go
RUN go build -o ws_client client/main.go

FROM alpine:3.24 AS server
RUN apk add --update ca-certificates && rm -rf /var/cache/apk/*
RUN addgroup -S app && adduser -S -G app app
WORKDIR /app
COPY --from=build-base --chown=app:app /app/hello-websocket-go/ws_server /app/
USER app
ENTRYPOINT ["./ws_server"]

FROM alpine:3.24 AS client
RUN apk add --update ca-certificates && rm -rf /var/cache/apk/*
RUN addgroup -S app && adduser -S -G app app
WORKDIR /app
COPY --from=build-base --chown=app:app /app/hello-websocket-go/ws_client /app/
USER app
ENTRYPOINT ["./ws_client"]
